tensorboard/components/tensor_widget/slicing-control.ts

```diff
@@ -0,0 +1,335 @@
+   *   will be invoked for changes in:
+   *   - which dimensions are used for slicing and which for viewing
+   *   - the selected index within a slicing dimension.
+   */
+  constructor(
+    private readonly rootDiv: HTMLDivElement,
+    private readonly shape: Shape,
+    private readonly onSlicngSpecChange?: OnSlicingSpecChangeCallback
+  ) {
+    this.rank = this.shape.length;
+    if (this.rank < 3) {
+      throw new Error(
+        `Dimension control is not applicable to tensor shapes less than ` +
+          `3D: received ${this.rank}D tensor shape: ` +
+          `${JSON.stringify(this.shape)}.`
+      );
+    }
+    this.createComponents();
+  }
+
+  /**
+   * Create all UI components of this SlicingControl.
+   *
+   * The detailed contents of the components are not filled in. Those are filled
+   * in when the `render()` method is called.
+   */
+  private createComponents() {
+    // Clear the dim group.
+    while (this.rootDiv.firstChild) {
+      this.rootDiv.removeChild(this.rootDiv.firstChild);
+    }
+    this.dimControls = [];
+    this.dimInputs = [];
+    this.dropdowns = [];
+    this.dimControlsListenerAttached = [];
+
+    // Create the div elements for the brackets and the dim controls.
+    this.bracketDivs[0] = document.createElement('div');
+    this.bracketDivs[0].textContent = 'Slicing: [';
+    this.bracketDivs[0].classList.add('tensor-widget-dim-brackets');
+    this.rootDiv.appendChild(this.bracketDivs[0]);
+
+    for (let i = 0; i < this.rank; ++i) {
+      const dimControl = document.createElement('div');
+      dimControl.classList.add('tensor-widget-dim');
+      this.rootDiv.appendChild(dimControl);
+      this.dimControls.push(dimControl);
+
+      this.dimControlsListenerAttached.push(false);
+
+      const dimInput = document.createElement('input');
+      dimInput.classList.add('tensor-widget-dim');
+      // The dim input is initially hidden, and will be shown when the
+      // corresponding dim control is clicked.
+      dimInput.style.display = 'none';
+      this.rootDiv.appendChild(dimInput);
+      this.dimInputs.push(dimInput);
+
+      const dropdown = document.createElement('div');
+      dropdown.classList.add('tensor-widget-dim-dropdown');
+      // The dropdown is initially hidden, and will be shown when the
+      // corresponding dim control is clicked.
+      dropdown.style.display = 'none';
+      this.rootDiv.appendChild(dropdown);
+      this.dropdowns.push(dropdown);
+    }
+
+    this.bracketDivs[1] = document.createElement('div');
+    this.bracketDivs[1].textContent = ']';
+    this.bracketDivs[1].classList.add('tensor-widget-dim-brackets');
+    this.rootDiv.appendChild(this.bracketDivs[1]);
+  }
+
+  /**
+   * Re-render the slicing control according to the current slicing spec.
+   */
+  render(slicingSpec?: TensorViewSlicingSpec) {
+    if (slicingSpec != null) {
+      this.slicingSpec = JSON.parse(JSON.stringify(slicingSpec));
```

Comment:
This is a way to deep-copy the object. See the accepted answer in https://stackoverflow.com/questions/122102/what-is-the-most-efficient-way-to-deep-clone-an-object-in-javascript

We want to avoid mutating the input argument `slicingSpec`.

Reply:
Sure, but note jeresig’s comment and the actual benchmarks showing that
stringify–parse is (as should be expected) the _slowest_ way to copy:
<https://stackoverflow.com/a/5344074/>

We already have a lodash dependency; could you please use `_.cloneDeep`
instead?

Reply:
The object being deep-copied here is tiny. This method is not called frequently. If the performance becomes and issue in the future, we can consider using lodash. But for now, I think the benefit of one fewer dependency outweighs the tiny performance gain. (TensorWidget is intended to be usable as a standalone package)